lib/pbio/drv/legodev/legodev_pup.c

```diff
@@ -325,6 +330,13 @@ PROCESS_THREAD(pbio_legodev_pup_process, ev, data) {
             for (int i = 0; i < PBDRV_CONFIG_LEGODEV_PUP_NUM_EXT_DEV; i++) {
                 dev = &ext_devs[i];
 
+                // If UART device was connected but not reporting data, restart detection.
+                if (dev->connected_type_id == PBDRV_LEGODEV_TYPE_ID_LPF2_UNKNOWN_UART &&
+                    dev->connected_type_id == dev->prev_type_id && timer_expired(&dev->dcm.watchdog)) {
+                    dev->connected_type_id = PBDRV_LEGODEV_TYPE_ID_NONE;
+                    dev->dcm.dev_id_match_count = 0;
```

Comment:
Modifying the DCM state outside of the DCM module feels like spaghetti code. IMHO it would be better if UART sensor code could be reused independent of a DCM and it would be easier to understand the DCM if all state is managed in a single file.

Reply:
That's what this commit is working towards. Note that this piece of code was already here, just moved.

Reply:
> Modifying the DCM state outside of the DCM module feels like spaghetti code.

So this reduces it. It used to be modified from outside the DCM, and not anymore after this change. This is why it could get stuck before and not anymore now.

Reply:
I've started gathering some ideas to further separate them in https://github.com/pybricks/support/issues/1140